docs/architecture/adr-014-improve-permissions.md

```diff
@@ -0,0 +1,202 @@
+# ADR 014: Improve the permission system
+
+## Changelog
+
+- May 19th, 2022: First draft;
+
+## Status
+
+DRAFT Not Implemented
+
+## Abstract
+
+This ADR introduces a new way of registering and managing permissions within the `x/subspaces` module all the other modules that depend on it.
+
+## Context
+
+Currently, Desmos implements a permission system that is very similar to the one used within Linux-based systems: each permission is represented by an `uint` value treated reading its bits individually. The combination of a permission is thus obtained using the `|` operator, and to check if a specific permission is set it's sufficient to check the value of the associated bit. Although this works, it has some limitations to it:
+- clients need to know the value of each permission, and how to properly combine them to form the final permission value they want to set;
+- permission values that are given from outside (e.g. a client setting a permission for a user) need to be sanitized in order to grant forward-compatibility;
+- all permissions need to be put inside the `x/subspaces` module, so that a proper validity mask can be computed and used to sanitize permissions values.
+
+## Decision
+
+To allow for an easier way of managing permissions, we will change the entire permission system from being based on a bitwise system to be based on a more human-readable system. Instead of representing each permission value as a `uint` and then reading individual bits, we will represent each permission using a `string` value that can easily tell what the permission is about. Here are some examples:
+
+```go
+package types
+
+type Permission string
+
+const (
+	PermissionManageSubspace Permission = "PERMISSION_MANAGE_SUBSPACE"
+	PermissionEverything     Permission = "PERMISSION_EVERYTHING"
+)
+```
+
+In order to properly support any kind of permission from different modules, we will implement the following method within the `x/subspaces` module:
+
+```go
+var (
+	// registeredPermissions represents the list of permissions that are registered and should be considered valid
+	registeredPermissions []Permission
+)
+
+// RegisterPermission allows to register the permission with the given name and returns its value
+func RegisterPermission(permissionName string) Permission {
+	permission := Permission(strings.ToUpper(strings.ReplaceAll(permissionName, " ", "_")))
```

Comment:
Since users will be one day able to create their own, we should take under consideration the fact that `permissionName`s could arrive here in different formats, e.g. : `Permission.Manage.Subspace` or `Permission12!_Manage_Subspace`. 
To avoid this, I think we can rely on two possible regular expressions to make sure inserted permissions:
1. Only contains letters and spaces: `/^[a-zA-Z\s]*$/` no number or special symbols).
2. Are in `camelCase` format: `(?<=[a-z])(?=[A-Z])|(?<=[A-Z])(?=[A-Z][a-z])`

Reply:
I don't think we should have anything like this. Ideally, in the future, the only users that will be able to specify custom permissions will be subspaces owners or admins. It's going to be their responsibilities to create permissions with a readable name. Otherwise it's going to be something that makes their lives harder. It won't anyway impact any other user or subspace owner as permissions will be stored within each subspace anyway.

Reply:
Ok fine. Then I guess we're fine like this!